appsec/src/helper/remote_config/client.cpp

```diff
@@ -6,240 +6,186 @@
+        for (const product p : listener->get_supported_products()) {
+            std::vector<listener_base *> &vec_listeners =
+                listeners_per_product_[p];
+            vec_listeners.push_back(listener.get());
         }
     }
 }
 
-client::ptr client::from_settings(service_identifier &&sid,
+std::unique_ptr<client> client::from_settings(
     const remote_config::settings &settings,
-    std::vector<listener_base::shared_ptr> listeners)
+    std::vector<std::shared_ptr<listener_base>> listeners)
 {
-    return std::make_unique<client>(std::make_unique<http_api>(settings.host,
-                                        std::to_string(settings.port)),
-        std::move(sid), settings, std::move(listeners));
+    return std::unique_ptr<client>{new client(settings, std::move(listeners))};
 }
 
-[[nodiscard]] protocol::get_configs_request client::generate_request() const
+bool client::poll()
 {
-    std::vector<protocol::config_state> config_states;
-    std::vector<protocol::cached_target_files> files;
-
-    for (const auto &[product_name, product] : products_) {
-        // State
-        const auto configs_on_product = product.get_configs();
-        for (const auto &[id, config] : configs_on_product) {
-            config_states.push_back({config.id, config.version, config.product,
-                config.apply_state, config.apply_error});
-
-            std::vector<protocol::cached_target_files_hash> hashes;
-            hashes.reserve(config.hashes.size());
-            for (auto const &[algo, hash_sting] : config.hashes) {
-                hashes.push_back({algo, hash_sting});
-            }
-            files.push_back({config.path, config.length, std::move(hashes)});
-        }
+    SPDLOG_DEBUG("Polling remote config");
+
+    ddog_CharSlice slice{};
+    const bool has_update = ddog_remote_config_read(reader_.get(), &slice);
+    if (!has_update) {
+        SPDLOG_DEBUG("No update available for {}", settings_.shmem_path);
+        return false;
     }
 
-    const protocol::client_tracer ct{std::move(ids_.get()), sid_.tracer_version,
-        sid_.service, sid_.extra_services, sid_.env, sid_.app_version};
+    std::set<config> new_configs;
+    // NOLINTNEXTLINE
+    std::string_view resp{reinterpret_cast<const char *>(slice.ptr), slice.len};
+    auto pos_lf = resp.find('\n');
+    if (pos_lf == std::string_view::npos) {
+        throw std::runtime_error{
+            "Invalid response from remote config (no newline)"};
+        return false;
+    }
+    SPDLOG_DEBUG("Runtime id is {}", resp.substr(0, pos_lf));
 
-    const protocol::client_state cs{targets_version_, config_states,
-        !last_poll_error_.empty(), last_poll_error_, opaque_backend_state_};
-    std::vector<std::string> products_str;
-    products_str.reserve(products_.size());
-    for (const auto &[product_name, product] : products_) {
-        products_str.push_back(product_name);
+    std::string_view configs = resp.substr(pos_lf + 1);
```

Comment:
This will throw `std::out_of_range` if `pos_lf + 1` is beyond `size()`, although that might be the intended behaviour.

Reply:
we know that `pos_lf` is valid, so `substr(pos_lf +1)` can't be out of bounds ([example](https://godbolt.org/z/K4PqP3he7))